Support order creation notes

ThomazFB opened the pull request:
# Why 
Fix issue #6386 by defining a customer note section inside the New Order view capable of creating, editing or removing any text as the Order customer note field.

# How

Defines the `CustomerOrderNoteSection` inside the `NewOrder` view and handles all user inputs inside the `NewOrderViewModel` taking advantage of the already existing `EditCustomerNote` view.

| New Order view with the empty Customer note section  | Editing the customer note | New Order view with an existing Customer note input | Order created with a customer note |
| ------------- | ------------- | ------------- | ------------- |
| ![Simulator Screen Shot - iPhone 13 Pro Max - 2022-03-16 at 19 07 58](https://user-images.githubusercontent.com/5920403/158699477-7e075719-da78-449d-82d3-629404fa0047.png) | ![Simulator Screen Shot - iPhone 13 Pro Max - 2022-03-16 at 19 08 06](https://user-images.githubusercontent.com/5920403/158699488-3842fe81-5da2-45e8-8a42-96bbb7f57846.png) | ![Simulator Screen Shot - iPhone 13 Pro Max - 2022-03-16 at 19 08 08](https://user-images.githubusercontent.com/5920403/158699503-656e02c1-3f54-4d49-ac65-e7ef7a817f69.png) | ![Simulator Screen Shot - iPhone 13 Pro Max - 2022-03-16 at 19 08 19](https://user-images.githubusercontent.com/5920403/158699512-146ddf1f-c5e0-4991-9a83-e60ae1d7a38e.png) |

# Testing Steps

- Make sure the Order Creation experimental toggle is active inside the app settings
- Go to the Order list, hit the `+` icon and select `Order Creation`
- Verify the `Customer note` section, interact with it by adding, editing, or removing notes from it
- Make sure the discard warning works as expected with the customer note presence or not

 
Update release notes:
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

ThomazFB committed:
Create order notes basic file structure

ThomazFB committed:
Connect OrderNoteSection with NewOrderViewModel data

ThomazFB committed:
Define Notes SwiftUI preview

ThomazFB committed:
Add notes UI Layout and Localization constants

ThomazFB committed:
Create body structure for Notes section empty or with content

ThomazFB committed:
Create empty and filled notes view

ThomazFB committed:
Define edit button and add it to Notes content view

ThomazFB committed:
Readjust Notes section layout positioning

ThomazFB committed:
Add notes section to the New Order view

ThomazFB committed:
Update strings for Order notes

ThomazFB committed:
Define EditCustomerNoteViewModel creation inside NewOrderViewModel

ThomazFB committed:
Call EditCustomerNote view when `Add note` is clicked

ThomazFB committed:
Define basic OrderNotesViewModel implementation

ThomazFB committed:
Configure noteViewModel inside NewOrderViewModel

ThomazFB committed:
Configure EditCustomerNote to work directly from OrderNotesSection sheet call

ThomazFB committed:
Adjust synchronizers to account for notes

ThomazFB committed:
Adjust OrderNotesViewModel to work properly with EditCustomerNotes

ThomazFB committed:
Adjust data traffic between OrderNoteSection and NewOrderViewModel

ThomazFB committed:
Sync notes as a view state from OrderNotesSection

ThomazFB committed:
Update OrdersRemote to support customer notes on Order Creation

ThomazFB committed:
Update OrderStore.createOrder to include customerNotes field

ThomazFB committed:
Define OrderCreationNotesViewModel to allow code reuse for ViewModels

ThomazFB committed:
Refactor OrderNotesViewModel and SimplePaymentsNoteViewModel to avoid code repetition

ThomazFB committed:
Fix OrderStoreTests to account for customer_note as expected request key

ThomazFB committed:
Merge branch 'trunk' into issue/support-order-creation-notes

# Conflicts:
#	WooCommerce/Classes/ViewRelated/Orders/Order Creation/Synchronizer/LocalOrderSynchronizer.swift
#	WooCommerce/Classes/ViewRelated/Orders/Order Creation/Synchronizer/OrderSynchronizer.swift
#	WooCommerce/Classes/ViewRelated/Orders/Order Creation/Synchronizer/RemoteOrderSynchronizer.swift

ThomazFB committed:
Refactor Notes viewModels naming pattern

ThomazFB committed:
Adopt viewModel as the data structure for OrderNotesSection

ThomazFB committed:
Adjust OrderSynchronizer to work with Customer notes updates

ThomazFB committed:
Refactor naming system for Customer notes entities inside NewOrderViewModel

Ecarrion commented on Networking/Networking/Remote/OrdersRemote.swift:
```diff
@@ -323,5 +325,6 @@ public extension OrdersRemote {
         case billingAddress
         case shippingAddress
         case shippingLines
+        case customerNotes
```
```suggestion
        case customerNote
```

Ecarrion commented on WooCommerce/Classes/ViewRelated/Orders/Order Creation/NewOrderViewModel.swift:
```diff
@@ -240,6 +241,15 @@ final class NewOrderViewModel: ObservableObject {
     }
 
+    // MARK: Order notes data properties
+    @Published private(set) var customerNoteDataViewModel: CustomerNoteDataViewModel = .init(customerNotes: "")
+
+    lazy private(set) var noteViewModel = { OrderNotesViewModel(originalNote: customerNoteDataViewModel.customerNotes) }()
```
We try to have variables before the initializer and methods after, especially private ones who go on a private extension.

I dunno why `customerDataViewModel ` is there :-/

Ecarrion replied:
If this view model is our source of truth(holds the real data) I think it should be created with an empty string and not with the contents of `customerNoteDataViewModel` which I know also holds an empty string but makes the dependency relationship a bit confusing. 

What do you think?

ThomazFB replied:
I see! Makes perfect sense to me to define the properties before the init, I was trying to understand the class pattern and ended up mixing everything. About initializing the ViewModel with the empty string literal, I agree with your point. I've adjusted both suggestions! Thanks!

Ecarrion commented on WooCommerce/Classes/ViewRelated/Orders/Order Creation/NewOrderViewModel.swift:
```diff
@@ -406,6 +416,16 @@ extension NewOrderViewModel {
+    struct CustomerNoteDataViewModel {
+        let customerNotes: String
+
+        init(customerNotes: String) {
+            self.customerNotes = customerNotes
+        }
```
I think you don't need this initializer since it is automatically synthesized by swift for internal types!

ThomazFB replied:
Great to know that swift can do that! Nice one! Thanks!

Ecarrion commented on WooCommerce/Classes/ViewRelated/Orders/Order Creation/NotesSection/EditOrderNotesViewModel.swift:
```diff
@@ -0,0 +1,62 @@
+
+    /// Analytics engine.
+    ///
+    private let updateNoteAnalyticsTrackAction: () -> Void
```
I don't think this is needed! Since `newNote` is a published property, anyone can subscribe to it and perform any task, like updating analytics.

ThomazFB replied:
Indeed! I've removed this and went with a different strategy for the analytics trigger following how the `NewOrderViewModel` handles analytics. Let me know what you think about it.

Ecarrion commented on WooCommerce/Classes/ViewRelated/Orders/Order Creation/NotesSection/NewOrderCustomerNotesViewModel.swift:
```diff
@@ -0,0 +1,12 @@
+import Foundation
+import Combine
+
+final class OrderNotesViewModel: OrderCreationNotesViewModel {
+    init(originalNote: String = "", analytics: Analytics = ServiceLocator.analytics) {
+        super.init(originalNote: originalNote) {
+
+        }
+    }
+
+
+}
```
I'm not sure if I understand the purpose of this subclass.

I'm thinking that we can rename `OrderCreationNotesViewModel ` to something like `LocalOrderNoteViewModel` and reuse it in most places.

or am I missing something? 😅

ThomazFB replied:
I completely removed this as we're leaving the Simple Payments notes refactoring to another PR and changing how to trigger the Customer note analytics call, so now I defined a specific ViewModel for the New Order notes editing, it introduces some code repetition but it can be fixed in another PR as you suggested!

Ecarrion commented on WooCommerce/Classes/ViewRelated/Orders/Order Creation/NotesSection/OrderNotesSection.swift:
```diff
@@ -0,0 +1,126 @@
+            }.padding([.leading, .top, .trailing])
+
+            if viewModel.customerNotes.isEmpty {
+                createOrderNotesView
+            } else {
+                createNoteDataView
```
Nit: When I see the verb "create" I immediately think that this should be a function because is performing some operations. In contrast, computed variables are just accessing data. 

😅

Ecarrion replied:
What I mean is that I think this should be `createOrderNotesView()` and `createNoteDataView()` 😅

ThomazFB replied:
Makes perfect sense! I'm still wrapping my head about how SwiftUI perfectly supports what's expected from a Swift code, it's fixed now! Thanks!

Ecarrion commented on WooCommerce/Classes/ViewRelated/Orders/Order Creation/Synchronizer/OrderSynchronizer.swift:
```diff
@@ -81,6 +81,8 @@ protocol OrderSynchronizer {
     ///
     var setFee: PassthroughSubject<OrderFeeLine?, Never> { get }
 
+    var setNotes: PassthroughSubject<String?, Never> { get }
```
```suggestion
    var setNote: PassthroughSubject<String?, Never> { get }
```

Ecarrion commented on WooCommerce/Classes/ViewRelated/Orders/Order Creation/Synchronizer/RemoteOrderSynchronizer.swift:
```diff
@@ -143,6 +145,10 @@ private extension RemoteOrderSynchronizer {
             .assign(to: &$order)
+
+        setNotes.withLatestFrom(orderPublisher)
+            .map { notes, order in order.copy(customerNote: notes) }
+            .assign(to: &$order)
```
Small style nit:

```suggestion
            .map { notes, order  in 
                order.copy(customerNote: notes) 
            }
            .assign(to: &$order)
```

Ecarrion commented on WooCommerce/Classes/ViewRelated/Orders/Simple Payments/Summary/SimplePaymentsNoteViewModel.swift:
```diff
@@ -1,63 +1,10 @@
-    ///
-    private let analytics: Analytics
-
+class SimplePaymentsNoteViewModel: OrderCreationNotesViewModel {
```
I would suggest making the simple payments changes in a different PR! Even if they are small it helps a lot to the reviewers to focus their testing efforts in just one area!

ThomazFB replied:
Agreed! I've reverted all changes related to SimplePayments refactoring and I'm now leaving the Order Creation Customer note view model as a dedicated one with no code reuse, this way we can adjust everything in a different PR. Thanks!

ThomazFB committed:
Revert SimplePaymentsNoteViewModel to keep its original implementation

ThomazFB committed:
Update WooCommerce/Classes/ViewRelated/Orders/Order Creation/Synchronizer/OrderSynchronizer.swift

Co-authored-by: Ernesto Carrión <carrion.silver@gmail.com>

ThomazFB committed:
Update WooCommerce/Classes/ViewRelated/Orders/Order Creation/Synchronizer/RemoteOrderSynchronizer.swift

Co-authored-by: Ernesto Carrión <carrion.silver@gmail.com>

ThomazFB committed:
Update Networking/Networking/Remote/OrdersRemote.swift

Co-authored-by: Ernesto Carrión <carrion.silver@gmail.com>

ThomazFB committed:
Fix trailing whitespace violation

ThomazFB committed:
Make SimplePaymentsNoteViewModel final again

ThomazFB committed:
Rename setNotes to setNote inside OrderSynchronizer and respective implementations

ThomazFB committed:
Rename setNotes to setNote inside OrderSynchronizer and respective implementations

ThomazFB committed:
Refactor OrderNotesSection renaming it to CustomerOrderNoteSection

ThomazFB committed:
Fix customerNote enum reference in OrdersRemote

ThomazFB committed:
Fix setNote reference in NewOrderViewModel

ThomazFB committed:
Fix customerNote reference in OrderStore

ThomazFB committed:
Delete OrderNotesViewModel

ThomazFB committed:
Remove tracks HOF from EditOrderNotesViewModel

ThomazFB committed:
Update NewOrderViewModel to use directly the EditOrderNotesViewModel

ThomazFB committed:
Update project.pbxproj

ThomazFB committed:
Add Analytics track event for Order creation Customer note

ThomazFB committed:
Trigger customer note tracks for Order creation when updating the notes

ThomazFB committed:
Rename EditOrderNotesViewModel to CustomerOrderNoteViewModel

ThomazFB committed:
Reorganize some parts of documentation

ThomazFB committed:
Reorganize some parts of the NewOrderViewModel properties

ThomazFB committed:
Remove redundant init for CustomerNoteDataViewModel

ThomazFB committed:
Merge branch 'trunk' into issue/support-order-creation-notes

ThomazFB committed:
Make customer note changes trigger Discard warning

ThomazFB committed:
Create test coverage for customer note changes detection

ThomazFB committed:
Extract customer note track event to separate method

ThomazFB committed:
Create test coverage for customer note track events

ThomazFB committed:
Create test coverage for customerNote updates inside UI section

ThomazFB committed:
Refactor customer update action inside NewOrderViewModel

ThomazFB committed:
Update CustomerOrderNoteSection Localization strings

ThomazFB committed:
Update customer note string to match case pattern

ThomazFB commented:
@Ecarrion thank you for this great detailed and careful PR review! I've addressed all your suggestions, added the analytics track event for the Customer note, and added test coverage to the changes introduced here, so I'm now removing this PR as a draft and placing it for the complete review. Let me know what you think and what needs improvement! 🙇

ThomazFB committed:
Update fake order to contain empty customerNote instead of nil

ThomazFB committed:
Revert all changes in SimplePaymentsNoteViewModel

ThomazFB committed:
Refactor Customer note classes naming pattern

Ecarrion reviewed (commented):
This looks amazing @ThomazFB very good job! 🎉 

There is only one minor bug which is that we are not updating the local order with the local note changes after we come back from an order update.

You can see here how we are only setting back the local order status:

**Order Creation**
https://github.com/woocommerce/woocommerce-ios/blob/df4d399eddb86f31d622c15ebf8f81b0d3bd1ebc/WooCommerce/Classes/ViewRelated/Orders/Order%20Creation/Synchronizer/RemoteOrderSynchronizer.swift#L239-L242

**Order Update**
https://github.com/woocommerce/woocommerce-ios/blob/df4d399eddb86f31d622c15ebf8f81b0d3bd1ebc/WooCommerce/Classes/ViewRelated/Orders/Order%20Creation/Synchronizer/RemoteOrderSynchronizer.swift#L267-L269

You can see in this video the bug where after an order update (because we modified a product quantity) the order ends up with the old customer note(from when the order was created)

https://user-images.githubusercontent.com/562080/158833755-c691c946-ed6a-4f14-a6de-f9d4163f2689.mov

I think we could create a method like `updateOrderLocalState` where we set the order note & order status after an order completes creating/updating. 

What do you think?

ThomazFB committed:
Update RemoteOrderSynchronizer to update remote order with local customer note input

ThomazFB committed:
Create test coverage for customer note updates inside RemoteOrderSynchronizerTests

ThomazFB committed:
Create updateOrderWithLocalState extension method

ThomazFB commented:
Hey @Ecarrion, great catch! I've applied your fix suggestion and added some unit tests to cover the scenario you reproduced, let me know what you think! Changes applied [here](https://github.com/woocommerce/woocommerce-ios/pull/6398/commits/94262105f35e53131af5171941cfb834dd6ef8d0) and [here](https://github.com/woocommerce/woocommerce-ios/pull/6398/commits/ff8b3f78fadab477219e8700a149b57db3ef4096)

Ecarrion commented on WooCommerce/Classes/ViewRelated/Orders/Order Creation/Synchronizer/RemoteOrderSynchronizer.swift:
```diff
@@ -387,4 +385,9 @@ private extension Order {
     }
+
+    /// Return the targeted order with the current selected state.
+    func updateOrderWithLocalState(localOrder: Order) -> Order {
+        self.copy(status: localOrder.status, customerNote: localOrder.customerNote)
+    }
```
Conceptual nit: I'm not sure if having this method as an extension of `Order` makes much sense, as the order itself should not know which properties from `localOrder` are local.

I think, conceptually, this method should exist withing `RemoteOrderSynchronizer`.

In practice, it is the same, so feel free to leave it. Sorry for the rambling 😛

ThomazFB replied:
That's a good take! Maybe I let my Kotlin habits take over my decision for this one. The reason why I choose to do like this was to make it easier to read who's the receiving Order from who's the Local one with the data source passed as parameter, and set as an extension since this should be something exclusive from the `RemoteOrderSynchronizer`. But I don't want to diverge from Swift standards of coding or the way the project is organized around extensions, do you see it's best to make it a direct `RemoteOrderSynchronizer` function with no extension usage? I can change it with no worries at all 😄

Ecarrion replied:
I personally would have added it in `RemoteOrderSynchronizer`, because I think the knowledge of what is local and what is not local resides in `RemoteOrderSynchronizer`. 

Another alternative could have been to expose a static method from the synchronizer so the order extension can ask which properties are the local ones, but that may be even more convoluted.

In any case, it is not a big deal, so it is up to you!

ThomazFB replied:
Makes perfect sense, Ernesto! Thank you for sharing your thoughts on this one! Here's the change: https://github.com/woocommerce/woocommerce-ios/pull/6398/commits/01682884ce8ebbfaeb85354b5f483c016ecdfdd7. Let me know what you think!

Ecarrion commented on WooCommerce/WooCommerceTests/ViewRelated/Orders/Order Creation/Synchronizer/RemoteOrderSynchronizerTests.swift:
```diff
@@ -351,6 +351,85 @@ class RemoteOrderSynchronizerTests: XCTestCase {
+            switch action {
+            case .createOrder(_, _, let completion):
+                completion(.success(.fake().copy(orderID: self.sampleOrderID, customerNote: randomNote)))
+            case .updateOrder(_, _, _, _):
+                XCTFail("Unexpected action: \(action)")
```
I think this can be removed, right?

Ecarrion reviewed (approved):
Works great! very good job @ThomazFB!

ThomazFB committed:
Remove unnecessary update handling from customer note creating scenario

ThomazFB committed:
move order local update function outside of Order extension scope
